### PR TITLE
Tac Test Coverage

### PR DESCRIPTION
**Pull request by 5A11**
This PR adds test coverage for Tac Control Skill.

The PR specifically has the following changes. The changes in the code are marked for convenience.

- `COUNTERPARTY_NAME` is changed to `COUNTERPARTY_ADDRESS`
- Tac protocol is updated to make `UNREGISTER` performative a valid reply to `REGISTER`. Currently, UNREGISTER cannot be used.

## Types of changes

What types of changes does your code introduce to agents-aea?
_Put an `x` in the boxes that apply_

- [ ] Bugfix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)

## Checklist

_Put an `x` in the boxes that apply._

- [x] I have read the [CONTRIBUTING](../CONTRIBUTING.md) doc
- [x] I am making a pull request against the `develop` branch (left side). Also you should start your branch off our `develop`.
- [x] Lint and unit tests pass locally with my changes
- [x] I have added tests that prove my fix is effective or that my feature works
- [ ] I have checked that code coverage does not decrease.
- [ ] I have checked that the documentation about the `aea cli` tool works
- [ ] I have added necessary documentation (if appropriate)
- [x] Any dependent changes have been merged and published in downstream modules

### CODE REVIEW

File: packages/fetchai/protocols/tac/README.md [hunk 1]
```diff
@@ -69,7 +69,7 @@ ct:ErrorCode: |
 ---
 initiation: [register]
 reply:
-  register: [tac_error, game_data, cancelled]
+  register: [tac_error, game_data, cancelled, unregister]
```
Comment (5A11) [inline packages/fetchai/protocols/tac/README.md:72]: Tac protocol update.

File: packages/fetchai/protocols/tac/dialogues.py [hunk 2]
```diff
@@ -55,6 +55,7 @@ class TacDialogue(Dialogue):
                 TacMessage.Performative.TAC_ERROR,
                 TacMessage.Performative.GAME_DATA,
                 TacMessage.Performative.CANCELLED,
+                TacMessage.Performative.UNREGISTER,
```
Comment (5A11) [inline packages/fetchai/protocols/tac/dialogues.py:58]: Tac protocol update.

File: packages/fetchai/protocols/tac/protocol.yaml [hunk 3]
```diff
@@ -10,8 +10,8 @@ fingerprint:
   README.md: QmPWA1Z1w2XnH4a5GAG19hEapnWZB1DTcFLwAyYYNSR4pQ
   __init__.py: QmSAC7PGra9fig8RhhF1j3XEVpgie9UZNNYPc2AB9Kx9xJ
   custom_types.py: QmXQATfnvuCpt4FicF4QcqCcLj9PQNsSHjCBvVQknWpyaN
-  dialogues.py: QmQvuivrjhVFu7pjSFfv6FrWwVeBC7p8Hm3P4gjCnVx5Ym
-  message.py: QmfWnCcHXoLHMqAgQJHj9yqU7rhyGpXtFG2jh8vtN3XMCv
+  dialogues.py: Qma5kXzq6ydkvwBZTpPrDEWkSr4DvZVRM9mqSVgvJhvEFY
```
Comment (5A11) [inline packages/fetchai/protocols/tac/protocol.yaml:13]: Tac protocol update.

File: tests/test_packages/test_skills/test_generic_buyer/test_dialogues.py [hunk 4]
```diff
@@ -26,7 +26,7 @@
 from aea.exceptions import AEAEnforceError
 from aea.helpers.transaction.base import Terms
 from aea.protocols.dialogue.base import DialogueLabel
-from aea.test_tools.test_skill import BaseSkillTestCase, COUNTERPARTY_NAME
+from aea.test_tools.test_skill import BaseSkillTestCase, COUNTERPARTY_ADDRESS
```
Comment (5A11) [inline tests/test_packages/test_skills/test_generic_buyer/test_dialogues.py:29]: `COUNTERPARTY_NAME` to `COUNTERPARTY_ADDRESS` in this file

File: tests/test_packages/test_skills/test_generic_buyer/test_handlers.py [hunk 5]
```diff
@@ -33,7 +33,7 @@
     TransactionDigest,
```
Comment (5A11) [inline tests/test_packages/test_skills/test_generic_buyer/test_handlers.py:33]: COUNTERPARTY_NAME to COUNTERPARTY_ADDRESS across this file

File: tests/test_packages/test_skills/test_generic_buyer/test_models.py [hunk 6]
```diff
@@ -23,7 +23,7 @@
 from aea.configurations.constants import DEFAULT_LEDGER
 from aea.helpers.search.models import Constraint, ConstraintType, Description, Query
 from aea.helpers.transaction.base import Terms
-from aea.test_tools.test_skill import BaseSkillTestCase, COUNTERPARTY_NAME
+from aea.test_tools.test_skill import BaseSkillTestCase, COUNTERPARTY_ADDRESS
```
Comment (5A11) [inline tests/test_packages/test_skills/test_generic_buyer/test_models.py:26]: COUNTERPARTY_NAME to COUNTERPARTY_ADDRESS across this file

File: tests/test_packages/test_skills/test_generic_seller/test_dialogues.py [hunk 7]
```diff
@@ -26,7 +26,7 @@
 from aea.exceptions import AEAEnforceError
```
Comment (5A11) [inline tests/test_packages/test_skills/test_generic_seller/test_dialogues.py:26]: COUNTERPARTY_NAME to COUNTERPARTY_ADDRESS across this file

File: tests/test_packages/test_skills/test_generic_seller/test_handlers.py [hunk 8]
```diff
@@ -29,7 +29,7 @@
 from aea.helpers.search.models import Description
 from aea.helpers.transaction.base import Terms, TransactionDigest, TransactionReceipt
 from aea.protocols.dialogue.base import DialogueMessage, Dialogues
-from aea.test_tools.test_skill import BaseSkillTestCase, COUNTERPARTY_NAME
+from aea.test_tools.test_skill import BaseSkillTestCase, COUNTERPARTY_ADDRESS
```
Comment (5A11) [inline tests/test_packages/test_skills/test_generic_seller/test_handlers.py:32]: COUNTERPARTY_NAME to COUNTERPARTY_ADDRESS across this file

File: tests/test_packages/test_skills/test_generic_seller/test_models.py [hunk 9]
```diff
@@ -32,7 +32,7 @@
     Query,
 )
 from aea.helpers.transaction.base import Terms
-from aea.test_tools.test_skill import BaseSkillTestCase, COUNTERPARTY_NAME
+from aea.test_tools.test_skill import BaseSkillTestCase, COUNTERPARTY_ADDRESS
```
Comment (5A11) [inline tests/test_packages/test_skills/test_generic_seller/test_models.py:35]: COUNTERPARTY_NAME to COUNTERPARTY_ADDRESS across this file

File: tests/test_packages/test_skills/test_tac_control/test_parameters.py [hunk 10]
```diff
@@ -0,0 +1,226 @@
...
+        )
+        mocked_now_time = datetime.datetime.strptime(
+            "01 01 2020  00:02", "%d %m %Y %H:%M"
+        )
+        datetime_mock = Mock(wraps=datetime.datetime)
+        datetime_mock.now.return_value = mocked_now_time
+
+        with patch("datetime.datetime", new=datetime_mock):
+            with patch.object(self.parameters.context.logger, "log") as mocked_logger:
+                Parameters(
```
Comment (DavidMinarsch) [inline tests/test_packages/test_skills/test_tac_control/test_parameters.py:87]: I'd define kwargs once in setup as a dict, then reuse. Makes it easier to read and see what;'s different across instantiations.

File: tests/test_packages/test_skills/test_tac_control/test_parameters.py [hunk 11]
```diff
@@ -0,0 +1,226 @@
...
+            "01 01 2020  00:02", "%d %m %Y %H:%M"
+        )
+        datetime_mock = Mock(wraps=datetime.datetime)
+        datetime_mock.now.return_value = mocked_now_time
+
+        with patch("datetime.datetime", new=datetime_mock):
+            with patch.object(self.parameters.context.logger, "log") as mocked_logger:
+                par = Parameters(
+                    ledger_id="",
+                    contract_address=None,
```
Comment (DavidMinarsch) [inline tests/test_packages/test_skills/test_tac_control/test_parameters.py:133]: see above

File: tests/test_packages/test_skills/test_tac_control/test_parameters.py [hunk 12]
```diff
@@ -0,0 +1,226 @@
...
+        with pytest.raises(
+            ValueError,
+            match="If the contract address is set, then good ids and currency id must be provided and consistent.",
+        ):
+            assert Parameters(
+                ledger_id="",
+                contract_address="some_contract_address",
+                good_ids=[],
+                currency_ids=[],
+                min_nb_agents=2,
```
Comment (DavidMinarsch) [inline tests/test_packages/test_skills/test_tac_control/test_parameters.py:208]: see above

File: tests/test_packages/test_skills/test_tac_control/test_parameters.py [hunk 13]
```diff
@@ -0,0 +1,226 @@
...
+        """Setup the test class."""
+        super().setup()
+        cls.parameters = cast(Parameters, cls._skill.skill_context.parameters)
+        cls.parameters_manual = Parameters(
+            ledger_id="some_ledger_id",
+            contract_address=None,
+            good_ids=[],
+            currency_ids=[],
+            min_nb_agents=2,
+            money_endowment=200,
```
Comment (DavidMinarsch) [inline tests/test_packages/test_skills/test_tac_control/test_parameters.py:54]: see below

File: tests/test_packages/test_skills/test_tac_control/test_helpers.py [hunk 14]
```diff
@@ -0,0 +1,61 @@
...
+class TestHelpers(BaseSkillTestCase):
+    """Test Helper module methods of tac control."""
+
+    path_to_skill = Path(ROOT_DIR, "packages", "fetchai", "skills", "tac_control")
+
+    @classmethod
+    def setup(cls):
+        """Setup the test class."""
+        super().setup()
+
```
Comment (DavidMinarsch) [inline tests/test_packages/test_skills/test_tac_control/test_helpers.py:45]: some positive tests can be added as well

File: tests/test_packages/test_skills/test_tac_control/test_behaviours.py [hunk 15]
```diff
@@ -0,0 +1,700 @@
...
+        self.parameters._end_time = mocked_end_time
+
+        datetime_mock = Mock(wraps=datetime.datetime)
+        datetime_mock.now.return_value = mocked_now_time
+
+        self.parameters._min_nb_agents = 2
+        agent_1_address = "agent_address_1"
+        agent_1_name = "agent_name_1"
+        agent_2_address = "agent_address_2"
+        agent_2_name = "agent_name_2"
```
Comment (DavidMinarsch) [inline tests/test_packages/test_skills/test_tac_control/test_behaviours.py:353]: some repeat variables across tests could be reused

**Review by DavidMinarsch — COMMENTED** [review]
LGTM, some minor comments. Also, one test is failing.

File: tests/test_packages/test_skills/test_tac_control/test_parameters.py (hunk 10 above)
Comment (5A11) [reply tests/test_packages/test_skills/test_tac_control/test_parameters.py:87]: Done

File: tests/test_packages/test_skills/test_tac_control/test_parameters.py (hunk 13 above)
Comment (5A11) [reply tests/test_packages/test_skills/test_tac_control/test_parameters.py:54]: Done

File: tests/test_packages/test_skills/test_tac_control/test_parameters.py (hunk 11 above)
Comment (5A11) [reply tests/test_packages/test_skills/test_tac_control/test_parameters.py:133]: Done

File: tests/test_packages/test_skills/test_tac_control/test_parameters.py (hunk 12 above)
Comment (5A11) [reply tests/test_packages/test_skills/test_tac_control/test_parameters.py:208]: Done

File: tests/test_packages/test_skills/test_tac_control/test_behaviours.py (hunk 15 above)
Comment (5A11) [reply tests/test_packages/test_skills/test_tac_control/test_behaviours.py:353]: Done

File: tests/test_packages/test_skills/test_tac_control/test_helpers.py (hunk 14 above)
Comment (5A11) [reply tests/test_packages/test_skills/test_tac_control/test_helpers.py:45]: Done

File: aea/test_tools/test_skill.py [hunk 16]
```diff
@@ -67,6 +67,12 @@ def get_message_from_outbox(self) -> Optional[Message]:
         envelope = self._multiplexer.out_queue.get_nowait()
         return envelope.message
 
+    def dismiss_messages_from_outbox(self, number: int = 1) -> None:
```
Comment (DavidMinarsch) [inline aea/test_tools/test_skill.py:70]: `drop` rather than `dismiss` is probably more explicit.

Comment (5A11) [reply aea/test_tools/test_skill.py]: Done

File: packages/fetchai/skills/tac_control/game.py [hunk 17]
```diff
@@ -441,7 +441,7 @@ def __eq__(self, other):
         """Compare to another object."""
         return (
             isinstance(other, Transaction)
-            and super.__eq__()
+            and super().__eq__(other)
```
Comment (5A11) [inline packages/fetchai/skills/tac_control/game.py:444]: small change needed for `game.py` module.

File: tests/test_packages/test_skills/test_tac_control/test_game.py [hunk 18]
```diff
@@ -0,0 +1,1366 @@
...
+        self.transaction_1._sender_address = "some_sender_address"
+        self.transaction_1._counterparty_address = self.agent_address
+
+        # operation
+        self.agent_state.update(self.transaction_1)
+
+        # after
+        assert self.agent_state.amount_by_currency_id == {
+            "1": 20
+        }  # shouldn't this be 0? game module line 611: should it be -=?
```
Comment (5A11) [inline tests/test_packages/test_skills/test_tac_control/test_game.py:671]: Note the comment.

File: tests/test_packages/test_skills/test_tac_control/test_game.py [hunk 19]
```diff
@@ -0,0 +1,1366 @@
...
+        self.transaction_2._sender_address = "some_sender_address"
+        self.transaction_2._counterparty_address = self.agent_address
+
+        # operation
+        self.agent_state.update(self.transaction_2)
+
+        # after
+        assert self.agent_state.amount_by_currency_id == {
+            "1": 1
+        }  # shouldn't this be 19? game module line 611: should it be -=?
```
Comment (5A11) [inline tests/test_packages/test_skills/test_tac_control/test_game.py:686]: Note the comment.

File: tests/test_packages/test_skills/test_tac_control/test_game.py [hunk 20]
```diff
@@ -0,0 +1,1366 @@
...
+            {"2": 0, "3": 2},
+            {"2": 1.0, "3": 1.5},
+        )
+        expected_agent_state_2 = AgentState(
+            agent_address_2,
+            {"1": 20},  # shouldn't this be 0? game module line 611: should it be -=?
+            {"1": 1.0},
+            {"2": 2, "3": 2},
+            {"2": 1.0, "3": 1.5},
+        )
```
Comment (5A11) [inline tests/test_packages/test_skills/test_tac_control/test_game.py:1275]: Note the comment.

File: packages/fetchai/skills/tac_control/game.py (hunk 17 above)
Comment (DavidMinarsch) [reply packages/fetchai/skills/tac_control/game.py:444]: Good!

**Review by DavidMinarsch — APPROVED** [review]
LGTM

**Pull request merged**